### PR TITLE
refactor: refine error message and test for FORMAT UPSERT

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Motivation: previously the error message is very silly:

1. It first errors with `INCLUDE KEY clause must be set`
2. After it's added, it errors with `Primary key must be specified`

Why can't we hint enough information in one shot?

For `CREATE SOURCE UPSERT`, it's even more silly: It will go through the same error above, then 3. tell you `Source does not support PRIMARY KEY constraint`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: e2e_test/source_inline/kafka/include_key_as.slt [hunk 1]
```diff
@@ -0,0 +1,284 @@
+control substitution on
```
Comment (xxchan) [inline e2e_test/source_inline/kafka/include_key_as.slt:1]: The changes to this test can be reviewed commit by commit

Comment (tabVersion) [reply e2e_test/source_inline/kafka/include_key_as.slt:1]: what does this command mean?

**Review by tabVersion — APPROVED** [review]
LGTM, thanks for refining the error msg

File: e2e_test/source_inline/kafka/include_key_as.slt (hunk 1 above)
Comment (xxchan) [reply e2e_test/source_inline/kafka/include_key_as.slt:1]: Replace `${}` to env vars in the slt.

To use `${RISEDEV_KAFKA_WITH_OPTIONS_COMMON}`, this needs to be opened.

File: src/frontend/src/handler/create_source.rs [hunk 2]
```diff
@@ -1364,9 +1379,25 @@ pub async fn bind_create_source(
...
+        match source_schema.format {
+            Format::Upsert => {
+                return Err(ErrorCode::BindError(format!(
+                    "can't CREATE SOURCE with FORMAT UPSERT\n\nHint: use CREATE TABLE instead\n\n{}",
+                    hint_upsert(&source_schema.row_encode)
+                ))
+                .into());
+            }
+            _ => {
+                // TODO: enhance error message for other formats
```
Comment (xiangjinwu) [inline src/frontend/src/handler/create_source.rs:1398]: Debezium (DebeziumMongo), Canal, Maxwell all require `table` rather than `source`.

Comment (xxchan) [reply src/frontend/src/handler/create_source.rs:1398]: Yeah, I thought only `PLAIN` is for `SOURCE`. Now I'm not sure when I found none/native. 🤪

File: src/frontend/src/handler/create_source.rs [hunk 3]
```diff
@@ -793,34 +805,36 @@ pub(crate) async fn bind_source_pk(
...
                 // so we don't ignore ascii case here
-                if !key_column_name.eq(sql_defined_pk_names[0].as_str()) {
+                if sql_defined_pk_names.len() != 1
+                    || !key_column_name.eq(sql_defined_pk_names[0].as_str())
+                {
                     return Err(RwError::from(ProtocolError(format!(
-                        "upsert {}'s key column {} not match with sql defined primary key {}",
-                        encode, key_column_name, sql_defined_pk_names[0]
+                        "Only {} can be used as primary key\n\n{}",
+                        sql_defined_pk_names[0],
```
Comment (xiangjinwu) [inline src/frontend/src/handler/create_source.rs:820]:
```suggestion
                        "Only {} can be used as primary key\n\n{}",
                        key_column_name,
```

File: src/frontend/src/handler/create_source.rs (hunk 2 above)
Comment (tabVersion) [reply src/frontend/src/handler/create_source.rs:1398]: `native` is for datagen, so it should work in both source and table. Not sure about `None` 😇

Comment (BugenZhao) [reply src/frontend/src/handler/create_source.rs:1398]: Link to the original PR: https://github.com/risingwavelabs/risingwave/pull/12306

TL;DR: Keywords like `NATIVE` will never be explicitly written by users in their SQL, but appear after normalization and persistence. When we're going to do schema change, we have to reparse the SQL (under current implementation) so `NATIVE` must be accepted as a keyword.

File: src/frontend/src/handler/create_source.rs [hunk 4]
```diff
@@ -750,6 +750,18 @@ pub(crate) fn bind_all_columns(
     }
 }
 
+fn hint_upsert(encode: &Encode) -> String {
+    format!(
+        r#"Hint: For FORMAT UPSERT ENCODE {encode:}, INCLUDE KEY must be specified and the key column must be used as primary key.
```
Comment (BugenZhao) [inline src/frontend/src/handler/create_source.rs:755]: How I wish `thiserror` provides something similar to this:

https://docs.rs/snafu/latest/snafu/derive.Snafu.html#providing-data-beyond-the-error-trait

**Review by BugenZhao — APPROVED** [review]
L🥰TM. The tests for error reporting is quite comprehensive.

File: src/frontend/src/handler/create_source.rs [hunk 5]
```diff
@@ -750,6 +750,18 @@ pub(crate) fn bind_all_columns(
     }
 }
 
+fn hint_upsert(encode: &Encode) -> String {
+    format!(
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/frontend/src/handler/create_source.rs]: Not possible: https://arc.net/l/quote/ycjvwmwu

**Review by xiangjinwu — APPROVED** [review]

**Review by st1page — APPROVED** [review]

**Review by stdrc — APPROVED** [review]
👏👏
